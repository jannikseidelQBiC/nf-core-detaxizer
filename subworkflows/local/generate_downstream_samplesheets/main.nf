//
// Subworkflow with functionality specific to the nf-core/createtaxdb pipeline
//

workflow SAMPLESHEET_TAXPROFILER{
    take:
    ch_reads

    main:
    ch_header  = Channel.empty()
    format     = 'csv' // most common format in nf-core
    format_sep = ','

    // Make your samplesheet channel construct here depending on your downstream
    ch_list_for_samplesheet = ch_reads
                                .map {
                                    meta, reads ->
                                        def out_path            = file(params.outdir).toString() + '/filter/filtered/'
                                        def sample              = meta.id
                                        def run_accession       = meta.id - "_longReads"
                                        def instrument_platform = !meta.long_reads ? "ILLUMINA" : "OXFORD_NANOPORE"
                                        def fastq_1             = meta.single_end  ?  out_path + reads.getName(): out_path + reads[0].getName()
                                        def fastq_2             = !meta.single_end ? out_path + reads[1].getName() : ""
                                        def fasta               = ""
                                    [ sample: sample, run_accession:run_accession, instrument_platform:instrument_platform, fastq_1:fastq_1, fastq_2:fastq_2, fasta:fasta ]
                                }
                                .tap{ ch_colnames } //ch_header exists

    channelToSamplesheet(ch_colnames, ch_list_for_samplesheet, format, format_sep)

}

workflow GENERATE_DOWNSTREAM_SAMPLESHEETS {
    take:
    ch_reads

    main:
    def downstreampipeline_names = params.generate_pipeline_samplesheets.split(",")

    if ( downstreampipeline_names.contains('taxprofiler')) {
        SAMPLESHEET_TAXPROFILER(ch_reads)
    } else {
        error "Unsupported downstream pipeline: ${params.generate_pipeline_samplesheets}"
    }
}

def channelToSamplesheet(ch_header,ch_list_for_samplesheet, format, format_sep) {
    // Constructs the header string and then the strings of each row, and
    // finally concatenates for saving.
    ch_header
        .first()
        .map{ it.keySet().join(format_sep) }
        .concat( ch_list_for_samplesheet.map{ it.values().join(format_sep) })
        .collectFile(
            name:"${params.outdir}/downstream_samplesheet/${params.generate_pipeline_samplesheets}.${format}",
            newLine: true,
            sort: false
        )
}
