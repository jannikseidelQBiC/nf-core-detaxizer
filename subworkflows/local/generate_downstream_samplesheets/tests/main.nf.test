nextflow_workflow {
    name "Test Subworkflow GENERATE_DOWNSTREAM_SAMPLESHEETS"
    script "../main.nf"
    workflow "GENERATE_DOWNSTREAM_SAMPLESHEETS"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/generate_downstream_samplesheets"

    test("reads - taxprofiler") {

        when {
            params {
                outdir                          = "$outputDir"
                generate_pipeline_samplesheets  = 'taxprofiler'
            }
            workflow {
                """
                input[0] = Channel.of(
                        [
                            [id:'test_paired-end_plus_long-reads_longReads', single_end:true, long_reads:true, amount_of_files:1],
                            file('/workspace/detaxizer/work/4c/9dd808a8f11e655c9002cf8cd18e96/test_paired-end_plus_long-reads_longReads_filtered.fastq.gz')
                        ],
                        [
                            [id:'test_single-end_long_longReads', single_end:true, long_reads:true, amount_of_files:1],
                            file('/workspace/detaxizer/work/77/93139d711cd3ff3f70aae9448abe1e/test_single-end_long_longReads_filtered.fastq.gz')
                        ],
                        [
                            [id:'test_single-end_short', single_end:true, long_reads:false, amount_of_files:1],
                            file('/workspace/detaxizer/work/46/60c8e50c0a9f7eeb3031f541d635ba/test_single-end_short_filtered.fastq.gz')
                        ],
                        [
                            [id:'test_paired-end_plus_long-reads', single_end:false, long_reads:false, amount_of_files:2],
                            [
                                file('/workspace/detaxizer/work/0e/4910ba06048aca8c83040c928b85a1/test_paired-end_plus_long-reads_R1_filtered.fastq.gz'),
                                file('/workspace/detaxizer/work/0e/4910ba06048aca8c83040c928b85a1/test_paired-end_plus_long-reads_R2_filtered.fastq.gz')
                            ]
                        ],
                        [
                            [id:'test_paired-end', single_end:false, long_reads:false, amount_of_files:2],
                            [
                                file('/workspace/detaxizer/work/a5/07e1debd2d2f159501460984e82fc5/test_paired-end_R1_filtered.fastq.gz'),
                                file('/workspace/detaxizer/work/a5/07e1debd2d2f159501460984e82fc5/test_paired-end_R2_filtered.fastq.gz')
                            ]
                        ]
                )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot("${params.outdir}/downstream_samplesheets/taxprofiler.csv").match() }
            )
        }
    }
}
