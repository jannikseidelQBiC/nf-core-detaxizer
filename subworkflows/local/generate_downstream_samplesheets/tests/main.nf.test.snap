{
    "reads - taxprofiler,mag": {
        "content": [
            [
                "taxprofiler.csv:md5,3af063f070342c72605b0277a7df4c36",
                "mag.csv:md5,cb28c6bbf1ff5028a6c31dcbbc5f815a"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-14T11:08:57.333343647"
    }
}